File: ruby/grade-school/_iterations/1/grade_school.rb
class School
  def initialize
    @grades = Hash.new { |h, g| h[g] = [] }
  end

  def grade(g)
    @grades[g].sort
  end

  def roster
    @grades.keys.sort.flat_map { |g| grade(g) }
  end

  def add(name, grade)
    return false if @grades.any? { |_, names| names.include? name }
    @grades[grade] << name
    true
  end
end

File: clojure/grade-school/_iterations/2/src/grade_school.clj
(ns grade-school)

(defn grade
  "Returns the students registered in the school roster in the given grade."
  [school grade]
  (school grade []))

(defn add
  "Adds a student to the school roster in the given grade."
  [school name grade]
  (merge-with into school {grade [name]}))

(defn sorted
  "Returns a list of all grades in the school roster.
   Grades are sorted in ascending order, and students are sorted alphabetically in a grade."
  [school]
  (->> school
       (mapv (fn [[grade names]] [grade (sort names)]))
       (into (sorted-map))))
